client/src/components/logViewer/logParser.ts

```diff
@@ -0,0 +1,438 @@
+// Copyright © 2023, SAS Institute Inc., Cary, NC, USA.  All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// It can't be shown from the definition, but actually LogType can be uppercase, partial uppercase, as well as lowercase
+// for example, 'ERROR' and 'Error' are acceptable and will be treated as 'error'
+type LogType = "error" | "warning" | "note" | "title" | "source" | "normal";
+
+type ProblemType = "error" | "warning";
+
+export type LogItem = {
+  line: string;
+  type: LogType;
+  [key: string]: number | string;
+};
+
+export type Problem = {
+  lineNumber: number;
+  startColumn: number;
+  endColumn: number;
+  message: string;
+  type: ProblemType;
+};
+
+type ProblemLocation = {
+  startColumn: number;
+  endColumn: number;
+  lineNumber: number;
+};
+
+// the start column for code in log may be different with which is in source file,
+// one case is the \t is converted to multiple spaces, so column offset is needed.
+// the line number is different with which is in source file as well, so line offset is needed.
+type LocationOffset = {
+  columnOffset: number;
+  lineOffset: number;
+};
+
+type ProblemMetadata = {
+  codeLogLine: string;
+  logLines: string[];
+  location: ProblemLocation;
+  // "   -----" {location indicator}
+  // "   22" {problem index at same location}
+  // "   76" {problem index at same location}
+  problemIndicsForSameLocation: string[];
+  type: ProblemType; //"error" | "warning";
+};
+
+export function parseLog(
+  logs: LogItem[],
+  firstCodeLine: string,
+): Problem[] | null {
+  if (logs.length === 0 || firstCodeLine.trim() === "") {
+    return null;
+  }
+
+  const latestLogs = filterOutTheLatestLogs(logs, firstCodeLine);
+
+  const problemRelatedLogs = filterOutProblemRelatedLogs(latestLogs);
+
+  const problems: Problem[] = [];
+  const problemMetadata: ProblemMetadata = {
+    codeLogLine: "",
+    logLines: [],
+    location: { startColumn: -1, endColumn: -1, lineNumber: -1 },
+    problemIndicsForSameLocation: [],
+    type: "error",
+  };
+  let offset: LocationOffset = { columnOffset: -1, lineOffset: -1 };
+  let currentCodeLogLine = "";
+
+  problemRelatedLogs.forEach((logItem) => {
+    if (isProblemTypeLog(logItem)) {
+      if (isProblemBeginningLogLine(logItem.line)) {
+        handleProblemMetadata(
```

Comment:
Here, we call `handleProblemMetadata`, which calls `addProblem`, which calls `problems.push`. This is a fair amount of indirection for mutating the `problems` array. I'm wondering if a better approach is that `handleProblemMetadata` _returns` a `Problem` that we can then push to the `problems` array, something like...
```js
problems.push(handleProblemMetadata(/*...*/))
```

UPDATE: I'm guessing you'll have to first check if a `problem` was returned (looking at `handleProblemMetadata` implementation)

Reply:
Renamed handleProblemMetadata() to addProblem(), original addProblem() function is removed.